### PR TITLE
feat(state): Add trie

### PR DESCRIPTION
**Pull request by yejiayu**
Dependent on [trie](https://github.com/paritytech/trie), a generic implementation of the Base-16 Modified Merkle Tree data structure.

To keep compatible with ETH and CITA-1.0, serialization still uses RLP.

cc @mohanson @u2

### CODE REVIEW

File: state/src/trie/codec.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        NULL_RLP.to_vec()
+    }
+
+    fn leaf_node(partial: &[u8], value: &[u8]) -> Vec<u8> {
+        let mut stream = RlpStream::new_list(2);
+        stream.append(&&*partial);
+        stream.append(&&*value);
+        stream.out()
+    }
+
```
Comment (u2) [inline state/src/trie/codec.rs]: Later I will review the left part.

**Comment by mohanson** [comment]
I will review the code tonight because I need time to understand the TRIE.

**Comment by janx** [comment]
> Base-16 Modified Merkle Tree

what's this? why do we need it?

**Comment by janx** [comment]
What is base-16 doing here? What's the different from MPT?

**Comment by yejiayu** [comment]
> What is base-16 doing here? What's the advantages over MPT?

`Base-16` is a 16 forks tree. 

No advantages. The implementation is consistent with `MPT` and just abstract out the KVDB and serialization

**Comment by yejiayu** [comment]
理解成 BASE16 也是对的，因为 node key 的取值范围确实是 BASE16(ASCII 0~F)，所以一个 branch node 才形成了 16 叉。

**Comment by rainchen** [comment]
> ohhhhhhhhhhhh... I thought it's BASE16 like BASE64.

I got the same feeling.


> Dependent on trie, a generic implementation of the Base-16 Modified Merkle Tree data structure.
> To keep compatible with ETH and CITA-1.0, serialization still uses RLP.

@yejiayu 
you should also add these comment into the codes, not just the PR description. It will help others to quick understand the codes without reading the whole PR discussions.

**Comment by yejiayu** [comment]
OK, I will comment in the new PR.

**Pull request merged**